New color scheme

klaukh opened the pull request:
This PR allows users to mix the global coloring and per-series colors.  Three notes of worth:
* The current API for the R package doesn't always maintain color order. When series are not modified in the order they appear in the underlying data, the colors assigned by the global coloring scheme get reordered (e.g., the last series modified will get the last color in the color vector created by the dygraphs lib.)  This PR does __not__ correct for that.
* The new coloring scheme has been updated for both `dySeries` and `dyGroup`, and I've tested with all the examples in the documentation for `dySeries`, `dyGroup`, and the `Plotters`; all dygraphs render appropriately on my end. 
* I'm having this weird thing where knitting `.Rmd` to `.html` eliminates the ability to zoom, though not in all dygraphs.  Sooooo, while I've changed the `http://rstudio.github.io/dygraphs` documentation for the `.Rmd` file, this PR does __not__ include the `.html` version. I need to rely on someone else to generate the `.html` that works properly. The package documentation has been updated accordingly.

pshevtsov commented:
@klaukh Could you please provide more details or examples, I can't spot any visual differences in dygraphs rendered from `master` and from your fork.

pshevtsov commented on R/series.R:
```diff
@@ -315,10 +308,55 @@ resolveStemPlot <- function(stemPlot, plotter) {
+`%||%` <- function(a,b){
+  if(!is.null(a)) return(a)
+  else return(b)
+}
```
No newline at the end of file.

klaukh commented:
Thanks for catching the new line at EOF

No visual differences was what I was hoping for.  The existing paradigm requires the user to enter a color for every series even if desiring for only a single series, enter the colors in exactly the right order, and cannot redefine colors without affecting the color of every other series.  With this PR I'm aiming to make it more flexible... users can define colors on a per series basis, redefine colors if so desired, while still mimicking the nice color scheme of the underlying JS lib.  Here are some examples

The original dygraph without changing colors (the JS lib handles the colors)
```{s}
lungDeaths <- cbind(fdeaths, mdeaths, ldeaths)
dygraph(lungDeaths)
```
![dygraph1](https://user-images.githubusercontent.com/21261860/32694279-4a9328b2-c70a-11e7-8a24-290ccc95e260.png)

Setting a single color:
```{s}
dygraph(lungDeaths) %>% dySeries('fdeaths', color = 'red')
```
![dygraph2](https://user-images.githubusercontent.com/21261860/32694295-8dd19672-c70a-11e7-97e5-3da2f49c4a21.png)

Setting, then redefining a single color:
```{s}
dygraph(lungDeaths) %>% 
  dySeries('fdeaths', color = 'red') %>% 
  dySeries('fdeaths', color = 'black')
```
![dygraph3](https://user-images.githubusercontent.com/21261860/32694297-95801be6-c70a-11e7-82ff-5b385c8f0460.png)

Setting, then redefining a single color, then setting a second color:
```{s}
dygraph(lungDeaths) %>% 
  dySeries('fdeaths', color = 'red') %>% 
  dySeries('fdeaths', color = 'black') %>% 
  dySeries('ldeaths', color = 'red')
```
![dygraph4](https://user-images.githubusercontent.com/21261860/32694299-9b25a94e-c70a-11e7-92a9-703e5a9a1d6b.png)

If you're good with the results, give the word... I'll update my PR fork to include all the changes to the main since I opened the PR, then send a note indicating it's ready for merging.

Hope this clarifies intent and results.

pshevtsov commented:
Thanks for clarifying @klaukh!

Looks good so far, I'm going to check it more carefully later today.

pshevtsov commented on R/group.R:
```diff
@@ -113,6 +110,18 @@ dyGroup <- function(dygraph,
   }
+  
+  # when setting up the first color, we start handling colors here 
+  if(!is.null(color) && is.null(dygraph$x$attrs$colors)){
```
Place a space before left parentheses, except in a function call ([Source](http://adv-r.had.co.nz/Style.html)). It should be
```r
if (!is.null(color) && is.null(dygraph$x$attrs$colors)) {
```

pshevtsov commented on R/group.R:
```diff
@@ -113,6 +110,18 @@ dyGroup <- function(dygraph,
+  
+  # when setting up the first color, we start handling colors here 
+  if(!is.null(color) && is.null(dygraph$x$attrs$colors)){
+      colors <- dygraphColors(dygraph, length(labels)-1)
```
Place spaces around all binary operators (=, +, -, <-, etc.) ([Source](https://google.github.io/styleguide/Rguide.xml#spacing)).

pshevtsov commented on R/group.R:
```diff
@@ -113,6 +110,18 @@ dyGroup <- function(dygraph,
+  }
+
+  # prepare the colors list for processing 
+  if(!is.null(dygraph$x$attrs$colors)) {
```
Put space after `if`.

pshevtsov commented on R/series.R:
```diff
@@ -89,6 +86,18 @@ dySeries <- function(dygraph,
   labels <- names(data)
   
+  # when setting up the first color, we start handling colors here 
+  if(!is.null(color) && is.null(dygraph$x$attrs$colors)){
```
Put space after `if` and before `{`

pshevtsov commented on R/series.R:
```diff
@@ -89,6 +86,18 @@ dySeries <- function(dygraph,
+  }
+
+  # prepare the colors list for processing 
+  if(!is.null(dygraph$x$attrs$colors)) {
```
Put space after `if`

pshevtsov commented on R/series.R:
```diff
@@ -315,10 +308,53 @@ resolveStemPlot <- function(stemPlot, plotter) {
-
+`%||%` <- function(a,b){
+  if(!is.null(a)) return(a)
+  else return(b)
```
No need for `else`.

klaukh commented:
Roger! Will adjust future PRs accordingly. Thx

klaukh commented:
And will change this one...

klaukh commented:
Adjusted as requested, plus a few more styling issues. Also, fixed #18, the single color bug. CRAN check successful except for the slider test. 

Will fix the example conflict and push updated commit later (had to step away from computer for a bit)
